rotkehlchen/db/eth2.py

```diff
@@ -366,29 +375,222 @@ def _validator_stats_process_queries(
+            )
+
+        from_ts_ms = ts_sec_to_ms(from_ts)
+        running_balances: dict[int, FVal] = defaultdict(lambda: ZERO)
+        for event in sorted(events, key=lambda x: x.timestamp):
+            if isinstance(event, EthDepositEvent):
+                running_balances[v_index := event.validator_index] += event.amount
+            elif isinstance(event, EthWithdrawalEvent):
+                v_index = event.validator_index
+                if event.is_exit_or_blocknumber is True:
+                    if event.timestamp >= from_ts_ms:  # only count pnl within the specified range
+                        exits_pnl[v_index] += event.amount - running_balances[v_index]
+                        running_balances[v_index] = ZERO
+                    else:
+                        continue
+                else:
+                    for request_event in withdrawal_request_events:
+                        if (
+                            request_event.extra_data is None or
+                            request_event.extra_data.get('validator_index') != v_index
+                        ):  # skip unrelated withdrawal requests
+                            continue
+                        if request_event.timestamp <= event.timestamp and request_event.amount == event.amount:  # has request - partial withdrawal  # noqa: E501
+                            running_balances[v_index] -= event.amount
+                            withdrawal_request_events.remove(request_event)  # remove this request event so we don't match it again.  # noqa: E501
+                            break
+                    else:  # skimming withdrawal - doesn't change the effective balance, but is counted as profit  # noqa: E501
+                        if event.timestamp >= from_ts_ms:  # only count pnl within the specified range  # noqa: E501
+                            withdrawals_pnl[v_index] += event.amount
+                        continue
+            elif (
+                event.event_type == HistoryEventType.INFORMATIONAL and
+                event.event_subtype == HistoryEventSubType.CONSOLIDATE and
+                event.extra_data is not None and
+                (source_validator := event.extra_data.get('source_validator_index')) is not None and  # noqa: E501
+                (target_validator := event.extra_data.get('target_validator_index')) is not None
+            ):
+                if target_validator not in validator_indices:
+                    continue
+
+                if source_validator not in balances_over_time and source_validator not in validator_indices:  # noqa: E501
+                    # also process the source validator's history so we know how much got consolidated  # noqa: E501
+                    new_balances_over_time, _, _ = self.process_validators_balances_and_pnl(
+                        from_ts=from_ts,
+                        to_ts=to_ts,
+                        validator_indices={source_validator},
+                    )
```

Comment:
this is a bit confusing. `process_accumulating_validators_balances_and_pnl` calls `process_validators_balances_and_pnl` that also calls `process_accumulating_validators_balances_and_pnl`. I'm not sure if you can end up in a bad recursion that goes out of hands or even if you are processing more than once the same events. Can this be simplified somehow? My biggest fear is that you end up processing multiple times the same events and that I don't know how many queries to the db you do + loops